### PR TITLE
Fixing Issues #593, #591, #565 & #556

### PR DESCRIPTION
**Pull request by RealTelefonmast**
Issue #593 "No Food" Thought/Warning:
- Thought is cached every 2000 ticks
- Food Source check now includes purchasing opportunities and not only the inventory
- New Purchase Opportunities:
  - Dispensers (Nutrient Paste, Replimat) can now be set to be a vending machine for guests, they will try to buy food from dispensers if a price is set, otherwise they will simply take it, collected silver can be withdrawn by haulers
  - Gastronomy Restaurants are considered as purchase opportunities and thus as a food source

Issue #591 Guests recruit prisoners to their faction:
(not fully tested but patch should apply)
- Guests now recruit prisoners to their host faction, meaning you receive the prisoners instead of them

Issue #565 & #556 Bedrooms need to be more attractive:
- Bed quality now considers facilities, each facility adding 10 points to the score
- Bed quality now considers Ideological preferences:
  - Pawns with "shared bed prohibited" will now avoid beds already owned by someone else, even when they would normally accept it
  - Preferable bed styles get up to 50 extra points
  - Required facilities by ideology (ie. sleep accelerators) add 10 points each

General Improvements:
- ThoughtWorkerCached, new base class for ThoughtWorkers, allowing to cache the ThoughtState and set a refresh interval
- Custom HospitalityConfigDef for xml-defined configuration of the mod
  - Currently supporting a list for patching dispenser-defs to be vending-machines
- FoodUtility class for general food-related helper methods
- HospitalityDefOf to cache defs
- HospitalityContent to cache Texture2D (potentially for materials/colors/other graphical static data)

Translation strings in Keyed Hospitality.xml where needed.

### CODE REVIEW

File: Source/Source/CompVendingMachine.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        public Building_NutrientPasteDispenser Dispenser => (Building_NutrientPasteDispenser) parent;
+
+        public int CurrentPrice
+        {
+            get => basePrice;
+            internal set => basePrice = Mathf.Clamp(value, 0, int.MaxValue);
+        }
+
+        public bool IsFree => CurrentPrice == 0;
+        public string TotalSold => $"${silverContainer.TotalStackCount}";
```
Comment (OrionFive) [inline Source/Source/CompVendingMachine.cs]: This should use the convert to currency function of RimWorld

File: Source/Source/Utilities/FoodUtility.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+                if (!vendingMachine.CanAffordFast(buyerGuest, out Thing silver)) return false;
+
+                vendingMachine.ReceivePayment(buyerGuest.inventory.innerContainer, silver);
+                return true;
+            }
+            return false;
+        }
+
+        public static bool WillConsume(Pawn pawn, ThingDef foodDef)
```
Comment (OrionFive) [inline Source/Source/Utilities/FoodUtility.cs:38]: Has this function been moved or is it double? It looks really familiar...

Comment (RealTelefonmast) [reply Source/Source/Utilities/FoodUtility.cs:38]: It's based of the Gastronomy.RestaurantStock.WillConsume() method, I found it while looking into how to make DiningSpots a viable option for available food sources. I adjusted it a bit as vending machines probably never have to dispense drugs.
Added a general check to consider restrictions too (even if they arent configurable for guests).

File: Source/Source/CompVendingMachine.cs (hunk 1 above)
Comment (RealTelefonmast) [reply Source/Source/CompVendingMachine.cs]: Resolved using the .ToStringMoney() extension.

**Pull request merged**